File: sdk/typescript/src/commands/memory.ts
import { Command } from 'commander';
import { PrismerClient } from '../index';

type ClientFactory = () => PrismerClient;

export function register(parent: Command, getIMClient: ClientFactory, _getAPIClient: ClientFactory): void {
  const mem = parent
    .command('memory')
    .description('Agent memory file management');

  // memory write --scope <scope> --path <path> --content <content>
  mem
    .command('write')
    .description('Write a memory file')
    .requiredOption('-s, --scope <scope>', 'memory scope')
    .requiredOption('-p, --path <path>', 'file path within scope')
    .requiredOption('-c, --content <content>', 'file content')
    .option('--json', 'output raw JSON response')
    .action(async (opts: { scope: string; path: string; content: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.memory.createFile({
          scope: opts.scope,
          path: opts.path,
          content: opts.content,
        });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const file = res.data;
        process.stdout.write(`Memory file created\n`);
        process.stdout.write(`  ID:    ${file.id}\n`);
        process.stdout.write(`  Scope: ${file.scope}\n`);
        process.stdout.write(`  Path:  ${file.path}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // memory read [file-id]
  mem
    .command('read [file-id]')
    .description('Read a memory file by ID, or filter by scope/path')
    .option('-s, --scope <scope>', 'filter by scope (used when no file-id given)')
    .option('-p, --path <path>', 'filter by path (used when no file-id given)')
    .option('--json', 'output raw JSON response')
    .action(async (fileId: string | undefined, opts: { scope?: string; path?: string; json: boolean }) => {
      const client = getIMClient();
      try {
        // Direct lookup by ID
        if (fileId) {
          const res = await client.im.memory.getFile(fileId);

          if (opts.json) {
            process.stdout.write(JSON.stringify(res, null, 2) + '\n');
            return;
          }

          if (!res.ok) {
            process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
            process.exit(1);
          }

          const file = res.data;
          process.stdout.write(`ID:    ${file.id}\n`);
          process.stdout.write(`Scope: ${file.scope}\n`);
          process.stdout.write(`Path:  ${file.path}\n`);
          process.stdout.write(`\n${file.content ?? ''}\n`);
          return;
        }

        // Filter by scope/path, auto-read if single match
        const listRes = await client.im.memory.listFiles({
          scope: opts.scope,
          path: opts.path,
        });

        if (opts.json) {
          if (listRes.ok && Array.isArray(listRes.data) && listRes.data.length === 1) {
            const detailRes = await client.im.memory.getFile(listRes.data[0].id);
            process.stdout.write(JSON.stringify(detailRes, null, 2) + '\n');
          } else {
            process.stdout.write(JSON.stringify(listRes, null, 2) + '\n');
          }
          return;
        }

        if (!listRes.ok) {
          process.stderr.write(`Error: ${listRes.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const files = listRes.data as Array<{ id: string; scope: string; path: string }>;

        if (files.length === 0) {
          process.stdout.write('No memory files found.\n');
          return;
        }

        if (files.length === 1) {
          // Auto-read single match
          const detailRes = await client.im.memory.getFile(files[0].id);
          if (!detailRes.ok) {
            process.stderr.write(`Error: ${detailRes.error?.message || 'Unknown error'}\n`);
            process.exit(1);
          }
          const file = detailRes.data;
          process.stdout.write(`ID:    ${file.id}\n`);
          process.stdout.write(`Scope: ${file.scope}\n`);
          process.stdout.write(`Path:  ${file.path}\n`);
          process.stdout.write(`\n${file.content ?? ''}\n`);
          return;
        }

        // Multiple matches — show table
        printFileTable(files);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // memory list
  mem
    .command('list')
    .description('List memory files')
    .option('-s, --scope <scope>', 'filter by scope')
    .option('--json', 'output raw JSON response')
    .action(async (opts: { scope?: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.memory.listFiles({ scope: opts.scope });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const files = res.data as Array<{ id: string; scope: string; path: string }>;

        if (files.length === 0) {
          process.stdout.write('No memory files found.\n');
          return;
        }

        printFileTable(files);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // memory delete <file-id>
  mem
    .command('delete <file-id>')
    .description('Delete a memory file by ID')
    .option('--json', 'output raw JSON response')
    .action(async (fileId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.memory.deleteFile(fileId);

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        process.stdout.write(`Deleted memory file: ${fileId}\n`);
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // memory compact <conversation-id>
  mem
    .command('compact <conversation-id>')
    .description('Create a compaction summary for a conversation')
    .option('--json', 'output raw JSON response')
    .action(async (conversationId: string, opts: { json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.memory.compact({ conversationId });

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const summary = res.data;
        process.stdout.write(`Compaction complete\n`);
        if (summary?.id) {
          process.stdout.write(`  Summary ID:      ${summary.id}\n`);
        }
        if (summary?.conversationId) {
          process.stdout.write(`  Conversation ID: ${summary.conversationId}\n`);
        }
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });

  // memory load
  mem
    .command('load')
    .description('Load session memory context')
    .option('-s, --scope <scope>', 'scope to load')
    .option('--json', 'output raw JSON response')
    .action(async (opts: { scope?: string; json: boolean }) => {
      const client = getIMClient();
      try {
        const res = await client.im.memory.load(opts.scope);

        if (opts.json) {
          process.stdout.write(JSON.stringify(res, null, 2) + '\n');
          return;
        }

        if (!res.ok) {
          process.stderr.write(`Error: ${res.error?.message || 'Unknown error'}\n`);
          process.exit(1);
        }

        const context = res.data;
        if (!context || (typeof context === 'object' && Object.keys(context).length === 0)) {
          process.stdout.write('No memory context available.\n');
          return;
        }

        process.stdout.write('Memory context loaded:\n\n');
        if (typeof context === 'string') {
          process.stdout.write(context + '\n');
        } else {
          process.stdout.write(JSON.stringify(context, null, 2) + '\n');
        }
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`Error: ${message}\n`);
        process.exit(1);
      }
    });
}

function printFileTable(files: Array<{ id: string; scope: string; path: string }>): void {
  const idLen = Math.max(2, ...files.map((f) => f.id.length));
  const scopeLen = Math.max(5, ...files.map((f) => f.scope.length));
  const pathLen = Math.max(4, ...files.map((f) => f.path.length));

  const row = (id: string, scope: string, path: string) =>
    `${id.padEnd(idLen)}  ${scope.padEnd(scopeLen)}  ${path.padEnd(pathLen)}`;

  process.stdout.write(row('ID', 'SCOPE', 'PATH') + '\n');
  process.stdout.write(`${'-'.repeat(idLen)}  ${'-'.repeat(scopeLen)}  ${'-'.repeat(pathLen)}\n`);
  for (const f of files) {
    process.stdout.write(row(f.id, f.scope, f.path) + '\n');
  }
}

File: sdk/prismer-cloud/golang/cmd/prismer/memory.go
package main

import (
	"fmt"

	prismer "github.com/Prismer-AI/PrismerCloud/sdk/prismer-cloud/golang"
	"github.com/spf13/cobra"
)

var memoryCmd = &cobra.Command{
	Use:   "memory",
	Short: "Agent memory file management",
}

// ── memory write ────────────────────────────────────────

var (
	memoryWriteScope   string
	memoryWritePath    string
	memoryWriteContent string
	memoryWriteJSON    bool
)

var memoryWriteCmd = &cobra.Command{
	Use:   "write",
	Short: "Write a memory file",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Memory.CreateFile(ctx, &prismer.CreateMemoryFileOptions{
			Scope:   memoryWriteScope,
			Path:    memoryWritePath,
			Content: memoryWriteContent,
		})
		if err != nil {
			return err
		}
		if memoryWriteJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Memory file created\n")
		fmt.Printf("  ID:    %v\n", data["id"])
		fmt.Printf("  Scope: %v\n", data["scope"])
		fmt.Printf("  Path:  %v\n", data["path"])
		return nil
	},
}

// ── memory read ─────────────────────────────────────────

var (
	memoryReadScope string
	memoryReadPath  string
	memoryReadJSON  bool
)

var memoryReadCmd = &cobra.Command{
	Use:   "read [file-id]",
	Short: "Read a memory file by ID, or filter by scope/path",
	Args:  cobra.MaximumNArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()

		if len(args) == 1 {
			// Direct lookup by ID
			res, err := client.Memory.GetFile(ctx, args[0])
			if err != nil {
				return err
			}
			if memoryReadJSON {
				return printJSON(res)
			}
			if !res.OK {
				return fmt.Errorf("API error: %v", res.Error)
			}
			data := asMap(res.Data)
			fmt.Printf("ID:    %v\nScope: %v\nPath:  %v\n\n%v\n",
				data["id"], data["scope"], data["path"], data["content"])
			return nil
		}

		// Filter by scope/path
		res, err := client.Memory.ListFiles(ctx, memoryReadScope, memoryReadPath)
		if err != nil {
			return err
		}
		if memoryReadJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		files := asList(res.Data)
		if len(files) == 0 {
			fmt.Println("No memory files found.")
			return nil
		}
		if len(files) == 1 {
			// Auto-read single match
			fm := asMap(files[0])
			id, _ := fm["id"].(string)
			if id != "" {
				detail, err := client.Memory.GetFile(ctx, id)
				if err == nil && detail.OK {
					data := asMap(detail.Data)
					fmt.Printf("ID:    %v\nScope: %v\nPath:  %v\n\n%v\n",
						data["id"], data["scope"], data["path"], data["content"])
					return nil
				}
			}
		}
		printMemoryTable(files)
		return nil
	},
}

// ── memory list ─────────────────────────────────────────

var (
	memoryListScope string
	memoryListJSON  bool
)

var memoryListCmd = &cobra.Command{
	Use:   "list",
	Short: "List memory files",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Memory.ListFiles(ctx, memoryListScope, "")
		if err != nil {
			return err
		}
		if memoryListJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		files := asList(res.Data)
		if len(files) == 0 {
			fmt.Println("No memory files found.")
			return nil
		}
		printMemoryTable(files)
		return nil
	},
}

// ── memory delete ───────────────────────────────────────

var memoryDeleteJSON bool
var memoryDeleteCmd = &cobra.Command{
	Use:   "delete <file-id>",
	Short: "Delete a memory file by ID",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Memory.DeleteFile(ctx, args[0])
		if err != nil {
			return err
		}
		if memoryDeleteJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		fmt.Printf("Deleted memory file: %s\n", args[0])
		return nil
	},
}

// ── memory compact ──────────────────────────────────────

var memoryCompactJSON bool
var memoryCompactCmd = &cobra.Command{
	Use:   "compact <conversation-id>",
	Short: "Create a compaction summary for a conversation",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Memory.Compact(ctx, &prismer.CompactOptions{
			ConversationId: args[0],
		})
		if err != nil {
			return err
		}
		if memoryCompactJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		data := asMap(res.Data)
		fmt.Printf("Compaction complete\n")
		if id, ok := data["id"]; ok {
			fmt.Printf("  Summary ID:      %v\n", id)
		}
		if convID, ok := data["conversationId"]; ok {
			fmt.Printf("  Conversation ID: %v\n", convID)
		}
		return nil
	},
}

// ── memory load ─────────────────────────────────────────

var (
	memoryLoadScope string
	memoryLoadJSON  bool
)

var memoryLoadCmd = &cobra.Command{
	Use:   "load",
	Short: "Load session memory context",
	RunE: func(cmd *cobra.Command, args []string) error {
		client := getIMClient().IM()
		ctx := cmd.Context()
		res, err := client.Memory.Load(ctx, memoryLoadScope)
		if err != nil {
			return err
		}
		if memoryLoadJSON {
			return printJSON(res)
		}
		if !res.OK {
			return fmt.Errorf("API error: %v", res.Error)
		}
		if res.Data == nil {
			fmt.Println("No memory context available.")
			return nil
		}
		fmt.Println("Memory context loaded:")
		fmt.Println()
		return printJSON(res.Data)
	},
}

// ── helpers ─────────────────────────────────────────────

func printMemoryTable(files []interface{}) {
	fmt.Printf("%-36s  %-20s  %s\n", "ID", "SCOPE", "PATH")
	fmt.Printf("%-36s  %-20s  %s\n", "------", "-----", "----")
	for _, f := range files {
		fm := asMap(f)
		fmt.Printf("%-36v  %-20v  %v\n", fm["id"], fm["scope"], fm["path"])
	}
}

func init() {
	memoryWriteCmd.Flags().StringVarP(&memoryWriteScope, "scope", "s", "", "Memory scope")
	memoryWriteCmd.Flags().StringVarP(&memoryWritePath, "path", "p", "", "File path within scope")
	memoryWriteCmd.Flags().StringVarP(&memoryWriteContent, "content", "c", "", "File content")
	memoryWriteCmd.Flags().BoolVar(&memoryWriteJSON, "json", false, "Output raw JSON")
	_ = memoryWriteCmd.MarkFlagRequired("scope")
	_ = memoryWriteCmd.MarkFlagRequired("path")
	_ = memoryWriteCmd.MarkFlagRequired("content")

	memoryReadCmd.Flags().StringVarP(&memoryReadScope, "scope", "s", "", "Filter by scope (used when no file-id given)")
	memoryReadCmd.Flags().StringVarP(&memoryReadPath, "path", "p", "", "Filter by path (used when no file-id given)")
	memoryReadCmd.Flags().BoolVar(&memoryReadJSON, "json", false, "Output raw JSON")

	memoryListCmd.Flags().StringVarP(&memoryListScope, "scope", "s", "", "Filter by scope")
	memoryListCmd.Flags().BoolVar(&memoryListJSON, "json", false, "Output raw JSON")

	memoryDeleteCmd.Flags().BoolVar(&memoryDeleteJSON, "json", false, "Output raw JSON")

	memoryCompactCmd.Flags().BoolVar(&memoryCompactJSON, "json", false, "Output raw JSON")

	memoryLoadCmd.Flags().StringVarP(&memoryLoadScope, "scope", "s", "", "Scope to load")
	memoryLoadCmd.Flags().BoolVar(&memoryLoadJSON, "json", false, "Output raw JSON")

	memoryCmd.AddCommand(memoryWriteCmd)
	memoryCmd.AddCommand(memoryReadCmd)
	memoryCmd.AddCommand(memoryListCmd)
	memoryCmd.AddCommand(memoryDeleteCmd)
	memoryCmd.AddCommand(memoryCompactCmd)
	memoryCmd.AddCommand(memoryLoadCmd)
	rootCmd.AddCommand(memoryCmd)
}
